File: JS/theShortestRoute/telegram.js
"use strict";

class Heap {
  constructor() {
    this.heap = [];
  }

  getLeftChildIndex = (parentIndex) => parentIndex * 2 + 1;
  getRightChildIndex = (parentIndex) => parentIndex * 2 + 2;
  getParentIndex = (childIndex) => Math.floor((childIndex - 1) / 2);

  insert = (key, value) => {
    const node = { key, value };
    this.heap.push(node);
    this.heapifyUp();
  };

  heapifyUp = () => {
    let index = this.heap.length - 1;
    const lastInsertedNode = this.heap[index];

    while (index > 0) {
      const parentIndex = this.getParentIndex(index);

      if (this.heap[parentIndex].key > lastInsertedNode.key) {
        this.heap[index] = this.heap[parentIndex];
        index = parentIndex;
      } else break;
    }

    this.heap[index] = lastInsertedNode;
  };

  remove = () => {
    const count = this.heap.length;
    const rootNode = this.heap[0];

    if (count <= 0) return undefined;
    else if (count === 1) {
      this.heap = [];
      return rootNode;
    } else {
      this.heap[0] = this.heap.pop();
      this.heapifyDown();

      return rootNode;
    }
  };

  heapifyDown = () => {
    let index = 0;
    const count = this.heap.length;
    const rootNode = this.heap[index];

    while (this.getLeftChildIndex(index) < count) {
      const leftChildIndex = this.getLeftChildIndex(index);
      const rightChildIndex = this.getRightChildIndex(index);

      const smallerChildIndex =
        rightChildIndex < count &&
        this.heap[rightChildIndex].key < this.heap[leftChildIndex].key
          ? rightChildIndex
          : leftChildIndex;

      if (this.heap[smallerChildIndex].key <= rootNode.key) {
        this.heap[index] = this.heap[smallerChildIndex];
        index = smallerChildIndex;
      } else break;
    }

    this.heap[index] = rootNode;
  };
}

class PriorityQueue extends Heap {
  constructor() {
    super();
  }

  enqueue = (priority, value) => this.insert(priority, value);
  dequeue = () => this.remove();
  isEmpty = () => this.heap.length <= 0;
}

const fs = require("fs");

const input =
  process.platform === "linux"
    ? fs.readFileSync("/dev/stdin").toString().split("\n")
    : `3 2 1
1 2 4
1 3 2`.split("\n");

const [N, M, start] = input[0].split(" ").map((v) => parseInt(v));
const graph = Array.from(Array(N + 1), () => Array());
const distance = Array(N + 1).fill(Infinity);

for (let i = 1; i < M + 1; i++) {
  const [a, b, c] = input[i].split(" ").map((v) => parseInt(v));

  graph[a].push({ node: b, distance: c });
}

const queue = new PriorityQueue();

queue.enqueue(0, start);
distance[start] = 0;

while (!queue.isEmpty()) {
  const { key, value } = queue.dequeue();
  const dist = key;
  const now = value;

  if (distance[now] < dist) {
    continue;
  }

  for (const i of graph[now]) {
    const cost = dist + i.distance;

    if (cost < distance[i.node]) {
      distance[i.node] = cost;
      queue.enqueue(cost, i.node);
    }
  }
}

let city = 0;
let time = 0;

for (let i = 1; i < N + 1; i++) {
  if (distance[i] !== Infinity && i !== start) {
    city++;
    time = Math.max(time, distance[i]);
  }
}

console.log(city, time);

File: src/theShortestRoute/Telegram.java
package theShortestRoute;

import java.io.*;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.PriorityQueue;

public class Telegram {
    static class Node{
        int node;
        int distance;

        public Node(int node, int dist) {
            this.node = node;
            this.distance = dist;
        }

        public int getNode() {
            return node;
        }

        public void setNode(int node) {
            this.node = node;
        }

        public int getDistance() {
            return distance;
        }

        public void setDistance(int distance) {
            this.distance = distance;
        }
    }

    static class queueNode implements Comparable<queueNode>{
        private int priority;
        private int value;

        public queueNode(int priority, int value) {
            this.priority = priority;
            this.value = value;
        }

        public int getPriority() {
            return priority;
        }

        public void setPriority(int priority) {
            this.priority = priority;
        }

        public int getValue() {
            return value;
        }

        public void setValue(int value) {
            this.value = value;
        }

        @Override
        public int compareTo(queueNode o) {
            if(this.priority > o.getPriority()){
                return 1;
            }else if(this.priority < o.getPriority()){
                return -1;
            }else{
                return 0;
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String[] nm = br.readLine().split(" ");
        int n = Integer.parseInt(nm[0]), m = Integer.parseInt(nm[1]), start = Integer.parseInt(nm[2]);

        List<List<Node>> graph = new ArrayList<>(n + 1);
        int[] distance =  new int[n + 1];

        for (int i = 0; i < n + 1; i++) {
            graph.add(new ArrayList<>());
            distance[i] = Integer.MAX_VALUE;
        }

        for (int i = 1; i < m + 1; i++) {
            String[] input = br.readLine().split(" ");
            int a = Integer.parseInt(input[0]), b = Integer.parseInt(input[1]), c = Integer.parseInt(input[2]);

            graph.get(a).add(new Node(b, c));
        }

        PriorityQueue<queueNode> queue = new PriorityQueue<>();
        queue.add(new queueNode(0, start));
        distance[start] = 0;

        while(!queue.isEmpty()){
            queueNode node = queue.remove();
            int dist = node.priority, now = node.value;

            if(distance[now] < dist){
                continue;
            }

            Iterator<Node> iterator = graph.get(now).iterator();

            while(iterator.hasNext()){
                Node next = iterator.next();
                int cost = dist + next.distance;

                if(cost < distance[next.node]){
                    distance[next.node] = cost;
                    queue.add(new queueNode(cost, next.node));
                }
            }
        }

        int city = 0, time = 0;

        for (int i = 1; i < n + 1; i++) {
            if(distance[i] != Integer.MAX_VALUE && i != start){
                city++;
                time = Math.max(time, distance[i]);
            }
        }

        bw.write(String.valueOf(city) + " " + String.valueOf(time));
        bw.flush();
        bw.close();
    }
}
